### PR TITLE
add `make_private_candidate_selection` (Rust)

### LINKED ISSUE #1781: Add private selection algorithm with known threshold
The private selection algorithm is a differentially private mechanism that repeatedly executes another DP mechanism until it achieves satisfactory performance according to a given quality score and threshold, incurring only a constant factor on the privacy loss budget. See Algorithm 1 in [Liu and Talwar (STOC, 2021)](https://arxiv.org/abs/1811.07971).

### PR DESCRIPTION
**Pull request by Shoeboxam**
Closes #1781

For PRB reviewers: feel free to just focus on the proof (.tex and pseudocode)!

### CODE REVIEW

**Comment by Shoeboxam** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/opendp/opendp/2126?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#2126** <a href="https://app.graphite.dev/github/pr/opendp/opendp/2126?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#1782** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1782?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Shoeboxam and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: rust/src/combinators/select_private_candidate/mod.rs [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+/// Algorithm 1 in `Private selection from private candidates <https://arxiv.org/pdf/1811.07971.pdf#page=7>`_ (Liu and Talwar, STOC 2019).
+///
+/// # Arguments
+/// * `measurement` - A measurement that releases a 2-tuple of (score, candidate)
+/// * `stop_probability` - The probability of stopping early at any iteration.
+/// * `threshold` - The threshold score. Return immediately if the score is above this threshold.
+/// * `max_iterations` - Optional. Run the algorithm at most this many iterations.
+///
+/// # Returns
+/// A measurement that returns a release from `measurement` whose score is greater than `threshold`, or none.
```
Comment (mccalluc) [inline rust/src/combinators/select_private_candidate/mod.rs:52]: Python usage example might be included? I guess I'm still leaning towards a little more documentation now, while it's on your mind.

File: rust/src/combinators/select_private_candidate/make_select_private_candidate.tex [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+\date{}
+
+\begin{document}
+
+\maketitle
+
+\contrib
+Proves soundness of \rustdoc{combinators/fn}{make\_select\_private\_candidate} in \asOfCommit{mod.rs}{f5bb719}.
+This constructor is a special case of \rustdoc{combinators/fn}{make\_select\_private\_candidate}.
+The proof for this constructor appeals to the proof for \rustdoc{combinators/fn}{make\_select\_private\_candidate}.
```
Comment (mccalluc) [inline rust/src/combinators/select_private_candidate/make_select_private_candidate.tex:15]: I'm confused here: It seems like the thing that you're proving is also what this is a special case of?

File: rust/src/combinators/select_private_candidate/references.bib [hunk 3]
```diff
@@ -0,0 +1,9 @@
+@misc{liu2018privateselectionprivatecandidates,
+      title={Private Selection from Private Candidates}, 
+      author={Jingcheng Liu and Kunal Talwar},
+      year={2018},
+      eprint={1811.07971},
+      archivePrefix={arXiv},
+      primaryClass={cs.DS},
+      url={https://arxiv.org/abs/1811.07971},
```
Comment (mccalluc) [inline rust/src/combinators/select_private_candidate/references.bib:8]: Surprised that this doesn't render as a link in the References section of the PDF?

**Review by mccalluc — APPROVED** [review]
A couple comments, and a couple places where I could be confused, but nothing that would block if it looks good to Shlomi.

File: rust/src/combinators/select_private_candidate/references.bib (hunk 3 above)
Comment (Shoeboxam) [reply rust/src/combinators/select_private_candidate/references.bib:8]: Agree, it's annoying. I'll make a pass through and update bib styling for everything at some point.

**Pull request merged**